### ISSUE TITLE
Clean code

### ISSUE THREAD

**Issue by stamnug** [taichi-dev/taichi#4498]
i'm learning the source and finding these code is ugly.  the reason is to a large degree on the limit of 80 char per line.
for example:
![QQ截图20220310022702](https://user-images.githubusercontent.com/21006005/157507012-4710a05c-14f8-49b5-b23d-222d8dc99488.png)
after modified:
![QQ截图20220310022634](https://user-images.githubusercontent.com/21006005/157507029-db875b2e-625a-4159-8dfa-587a6d802434.png)

but also there is some reason of style of writing.
for example:
![QQ截图20220310020833](https://user-images.githubusercontent.com/21006005/157504033-8e04bbb0-7129-4db6-96df-e6a7f006e063.png)
![QQ截图20220310021239](https://user-images.githubusercontent.com/21006005/157504723-606c7c53-2bed-4c28-9cb0-74448a8b6032.png)

they is very much and i don't know whether these modification is welcome.  if modifying them, I'm afraid you have a headache.  if can, hoping you are able to cost a little of time to typography. just like the code clean day of Blender.

**Comment by ailzhang** [comment]
Hey @stamnug , thanks a lot for the suggestion! FYI we use clang-format-10 with a slightly [customized config file ](https://github.com/taichi-dev/taichi/blob/master/.clang-format) to format our C++ code. We're happy to accept a PR to update clang-format config file to improve codebase as a whole instead of adhoc changes to a few files. 

Re writing style issues, small PRs to update them gradually are very welcome. We recommend opening an issue to discuss first before sending a giant PR just to give a heads up and avoid merge conflicts. 

Re code clean day, that's definitely a great idea! Btw we're actively working on enabling clang-tidy and cleaning up our codebase, any help on that is super welcome too!

**Comment by k-ye** [comment]
> the reason is to a large degree on the limit of 80 char per line.

Different ppl can have drastically different opinions on the coding style. For a project requiring multiple engineers to collaborate, the important thing is to stick to *one* style. It's like arguing whether vim or emacs is the best, which is mostly pointless because there is no real answer here.

---

The rest examples you pointed out are great points. It will be great to see the code getting improved, looking forward to your help!

**Comment by yihong0618** [comment]
@stamnug 
For code style issues, each project has its own compliance and author's preference, write in accordance with the style is good, also you can format it with any style you like in your `ENV` to read source code.

Suggestion, **ugly** that word is not very suitable for this aspect of the code format, there is a little hurt.

**Comment by stamnug** [comment]
瑕不掩瑜, 相对于代码的功能来讲, 代码的外观并不是那么重要. 如果有代码作者被我的提出的问题伤害到, 希望能有个正确的心态. 但问题终究还是要指出的, 出于尊重把不简洁的代码说成是个人喜好问题, 这对于开源项目并没有好处.  当然, Talk is cheap, ... , 所以代码作者完全没必要在意我的胡言乱语. 
关于代码格式问题, UE的代码真的非常漂亮, 漂亮到我不认为他们是用统一的工具进行格式化的, 我去研究研究他们怎么做的.

**Comment by ailzhang** [comment]
Hey @stamnug , we are super interested in improving the quality of our codebase. Although coding style issues sometimes get de-prioritized in our queue due to bandwidth limit, we welcome our community members to propose, contribute and bring the codebase to a better state together. 
If you find any best practice in the UE codebase that's portable to other codebase, we'd love to hear and learn from that! Thanks!

**Comment by stamnug** [comment]
> If you find any best practice in the UE codebase that's portable to other codebase, we'd love to hear and learn from that! Thanks!

![QQ图片20220310144710](https://user-images.githubusercontent.com/21006005/157604898-1de968f3-a19b-4e84-9a5a-30eeef1a40af.png)


其实我也不认为有工具能够真正格式化出美观整洁的代码.  所以我并不在工具上抱有期望. 或许你们可以fork一份UE代码参观一下. 
如果你们愿意花时间到人工修改代码上, 或许我能给出一点想法, 比如将代码整理的PR作为一个专项,  提交PR时进行标记:

#4498 * : 仅修改排版, 没有修改代码, 比如:
![QQ截图20220310151659](https://user-images.githubusercontent.com/21006005/157608873-3d4df579-4175-45d4-adb1-77a9e82cdbea.png)

#4498 ** : 修改了代码, 但没修改代码逻辑, 比如:
![QQ截图20220310151524](https://user-images.githubusercontent.com/21006005/157608691-cafb045a-1c8d-49c3-9781-6e495d3b3965.png)

#4498 *** : 修改了代码逻辑, 但没修改代码执行结果, 比如:
![QQ截图20220310151905](https://user-images.githubusercontent.com/21006005/157609129-97dc1a3a-37d4-4f06-affb-00c7c93269c9.png)

#4498 **** ban to define snode that len(axis) != len(shape), eg: ti.root.*.(ti.ij, 3) : 修改了执行结果, 并明确执行知道执行结果, 并在PR时指出. 比如:
![QQ截图20220310153502](https://user-images.githubusercontent.com/21006005/157612410-c89928eb-e194-4c93-a356-4e238a11aab1.png)


#4498 ***** :  可能导致未知的结果的修改.

不过如果有人确实认为函数参数每个参数一行比写一行更美观, 这确实是个问题, 还是有必要进行一次调研. 看看大家的意见.

**Comment by frostming** [comment]
There are two different topics in this issue, please don't mix them in discussion:

1. Code style, every project has its own standard, including max line length, brackets style, tab/space, etc. It may not fit your style, you can suggest but before that is landed, you have to follow since it is enforced by the CI.
2. Clean code, every suggestion to make the code cleaner and easier to read is appreciated. You can read the contributing guide and submit a PR, I believe the maintainers will be happy to review.

**Comment by stamnug** [comment]
我把标题改成了Clean code, 但我认为这是词不达意. 虽然ugly难听, 但这才是我的意思. 

每个人有自己的代码风格这没错,  比如不同命名方式就是不同风格, 比如花括号是单独一行还是放在行末也是不同风格.但代码是否美观, 易读, 我觉得和风格无关, 任何风格的代码都能写的美观, 同样也能写的混乱. 美观的代码也不会执着于一种风格. 而是会根据上下文选择合适的风格使代码更清晰. 但是, 我同样尊重任何程序员保持自己的风格. 我再次重申, 瑕不掩瑜. 何况这个瑕还是我的主观看法. 何况我一个动嘴皮子的, 有何资格说别人代码丑.

但是,我就是想表达一下我觉得源码丑这个观点, 丑多半是因为统一格式化的问题, 不知道为何会忽然转向到谈代码风格上, 我没有谈论任何程序员代码风格问题, 我只是说我觉得现在的代码丑. 如果没人和我有相同的看法, 那就当我没说好了.

**Comment by ailzhang** [comment]
> 其实我也不认为有工具能够真正格式化出美观整洁的代码. 所以我并不在工具上抱有期望. 或许你们可以fork一份UE代码参观一下.
如果你们愿意花时间到人工修改代码上, 或许我能给出一点想法, 比如将代码整理的PR作为一个专项, 提交PR时进行标记

IMHO automatic tooling is actually extremely powerful and helpful in an open source community as Taichi's. For anyone new to the codebase and community, it's the fastest way to collaborate without being blocked on stylish discrepancies. Given the limited dev bandwidth and high communication cost, it's a good trade-off to make. 

With that said, we are truly open to any concrete suggestions to the coding style and format config in the codebase. As long as it's backed by valid examples, we're happy to discuss, review and merge such PRs. 

@stamnug I'll go ahead and close this issue for now. Please feel free to open a new issue with if you have concrete suggestion to the formatting config. And we'll be happy to take PRs that improves the logic around `is_bit_level` and `self.mapping_key` mentioned above. Thanks a lot!
